### PR TITLE
Expand "use null propagation" feature to support additional common cases.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Teh analyzer/fixer now supports thsi pattern:

```c#
if (o != null)
    o.Something();

// converts to:
o?.Something();
```

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseNullPropagation/CSharpUseNullPropagationDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -36,29 +41,66 @@ protected override bool IsInExpressionTree(SemanticModel semanticModel, SyntaxNo
...
             isEquals = true;
 
             if (conditionNode is not IsPatternExpressionSyntax patternExpression)
-            {
                 return false;
-            }
 
-            if (patternExpression.Pattern is not ConstantPatternSyntax constantPattern)
+            var pattern = patternExpression.Pattern;
+            if (pattern is UnaryPatternSyntax(SyntaxKind.NotPattern) notPattern)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseNullPropagation/CSharpUseNullPropagationDiagnosticAnalyzer.cs:54]: this is new functionality for the existing analyzer.  we now detect `x is not null ? x.Y : null` which we didn't detect before.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -18,24 +18,40 @@ internal static class UseNullPropagationConstants
...
-        where TElementAccessExpression : TExpressionSyntax
+        where TInvocationExpressionSyntax : TExpressionSyntax
+        where TMemberAccessExpressionSyntax : TExpressionSyntax
+        where TConditionalAccessExpressionSyntax : TExpressionSyntax
+        where TElementAccessExpressionSyntax : TExpressionSyntax
+        where TIfStatementSyntax : TStatementSyntax
+        where TExpressionStatementSyntax : TStatementSyntax
     {
+        private static readonly ImmutableDictionary<string, string?> s_whenPartIsNullableProperties =
+            ImmutableDictionary<string, string?>.Empty.Add(UseNullPropagationConstants.WhenPartIsNullable, "");
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:53]: this already existed.  it was jus extracted from a local to a field.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -50,91 +66,74 @@ public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
...
 
+        protected abstract TSyntaxKind IfStatementSyntaxKind { get; }
         protected abstract ISyntaxFacts GetSyntaxFacts();
         protected abstract bool IsInExpressionTree(SemanticModel semanticModel, SyntaxNode node, INamedTypeSymbol? expressionTypeOpt, CancellationToken cancellationToken);
 
         protected abstract bool TryAnalyzePatternCondition(
-            ISyntaxFacts syntaxFacts, SyntaxNode conditionNode,
-            [NotNullWhen(true)] out SyntaxNode? conditionPartToCheck, out bool isEquals);
+            ISyntaxFacts syntaxFacts, TExpressionSyntax conditionNode,
+            [NotNullWhen(true)] out TExpressionSyntax? conditionPartToCheck, out bool isEquals);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:75]: a lot more members become more strongly typed to help avoid issues or catch them higher up.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -50,91 +66,74 @@ public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
...
-            whenFalseNode = syntaxFacts.WalkDownParentheses(whenFalseNode);
+            var conditionNode = (TExpressionSyntax)condition;
 
-            var conditionIsNegated = false;
-            if (syntaxFacts.IsLogicalNotExpression(conditionNode))
-            {
-                conditionIsNegated = true;
-                conditionNode = syntaxFacts.WalkDownParentheses(
-                    syntaxFacts.GetOperandOfPrefixUnaryExpression(conditionNode));
-            }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:112]: this removed check, and the removed check below both moved into TryAnalyzeCondition

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 5]
```diff
@@ -18,24 +18,40 @@ internal static class UseNullPropagationConstants
         public const string WhenPartIsNullable = nameof(WhenPartIsNullable);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:18]: this file is for the existing feature, and just represents cleanup to help make its logic more shareable.  The new functionality is added to a sibling part.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 6]
```diff
@@ -189,31 +184,41 @@ protected override void InitializeWorker(AnalysisContext context)
...
-                        context, syntaxFacts, referenceEqualsMethodOpt, invocation,
-                        out conditionPartToCheck, out isEquals);
-
-                default:
-                    return TryAnalyzePatternCondition(
-                        syntaxFacts, conditionNode, out conditionPartToCheck, out isEquals);
+                conditionIsNegated = true;
+                condition = (TExpressionSyntax)syntaxFacts.WalkDownParentheses(
+                    syntaxFacts.GetOperandOfPrefixUnaryExpression(condition));
             }
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:199]: this code moved in from above.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 7]
```diff
@@ -189,31 +184,41 @@ protected override void InitializeWorker(AnalysisContext context)
...
+
+                TInvocationExpressionSyntax invocation => TryAnalyzeInvocationCondition(
+                        context, syntaxFacts, referenceEqualsMethod, invocation,
+                        out conditionPartToCheck, out isEquals),
+
+                _ => TryAnalyzePatternCondition(syntaxFacts, condition, out conditionPartToCheck, out isEquals),
+            };
+
+            if (conditionIsNegated)
+                isEquals = !isEquals;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:214]: this code moved in from above.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 8]
```diff
@@ -226,22 +231,25 @@ protected override void InitializeWorker(AnalysisContext context)
...
+            IMethodSymbol? referenceEqualsMethod,
+            TInvocationExpressionSyntax invocation,
+            [NotNullWhen(true)] out TExpressionSyntax? conditionPartToCheck,
             out bool isEquals)
         {
             conditionPartToCheck = null;
             isEquals = true;
 
+            if (referenceEqualsMethod == null)
+                return false;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:251]: moved this check up from teh bottom so we don't do unnecessary work.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer_IfStatement.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer_IfStatement.cs:1]: this is the new functionality.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 10]
```diff
@@ -18,24 +18,40 @@ internal static class UseNullPropagationConstants
...
+    ///     x.Y();
+    /// </code>
+    /// </summary>
+    internal abstract partial class AbstractUseNullPropagationDiagnosticAnalyzer<
         TSyntaxKind,
         TExpressionSyntax,
+        TStatementSyntax,
         TConditionalExpressionSyntax,
         TBinaryExpressionSyntax,
-        TInvocationExpression,
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:26]: these were named incorrectly (missing hte 'Syntax' suffix).  Updated, which caused fallout below.

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 11]
```diff
@@ -50,104 +64,90 @@ public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
...
-            }
 
-            // ?. is not available in expression-trees.  Disallow the fix in that case.
+            // can't use ?. on a pointer
+            var whenPartType = semanticModel.GetTypeInfo(whenPartMatch, cancellationToken).Type;
+            if (whenPartType is IPointerTypeSymbol)
+                return;
 
-            var type = semanticModel.GetTypeInfo(conditionalExpression).Type;
+            // ?. is not available in expression-trees.  Disallow the fix in that case.
```
Comment (davidwengier) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:149]: Comment looks to be on the wrong line.. should be above line 164, no?

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer_IfStatement.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+        var semanticModel = context.SemanticModel;
+        var whenPartMatch = GetWhenPartMatch(
+            syntaxFacts, semanticModel, conditionPartToCheck,
+            (TExpressionSyntax)syntaxFacts.GetExpressionOfExpressionStatement(expressionStatement),
+            cancellationToken);
+        if (whenPartMatch == null)
+            return;
+
+        // can't use ?. on a pointer
```
Comment (davidwengier) [inline src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer_IfStatement.cs:62]: Does this method need the `Expression<T>` check too?

File: src/Analyzers/CSharp/Analyzers/UseNullPropagation/CSharpUseNullPropagationDiagnosticAnalyzer.cs [hunk 13]
```diff
@@ -36,29 +40,66 @@ protected override bool IsInExpressionTree(SemanticModel semanticModel, SyntaxNo
...
+            //
+            //   if (...)
+            //   {
+            //       statement
+            //   }
+
+            condition = ifStatement.Condition;
+
+            trueStatement = null;
+            if (ifStatement.Else == null)
```
Comment (davidwengier) [inline src/Analyzers/CSharp/Analyzers/UseNullPropagation/CSharpUseNullPropagationDiagnosticAnalyzer.cs:89]: Consider inverting this and returning early. The VB version of this was nicer to read IMO

**Review by davidwengier — APPROVED** [review]

File: src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer_IfStatement.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer_IfStatement.cs:62]: No. Statements can't appear in expression trees, so it's a non issue :-)

**Pull request merged**